### PR TITLE
Storage prefix configurable

### LINKED ISSUE #611: Storage file structure and S3 lifecycle transitions
Just a quick note—a flaw, maybe—on Storage:

I realized due to the directory structure `(public|private)/<userId>/...`, it's impossible to have an S3 lifecycle transition rule use a prefix. Typically I might set a prefix like `files/original/`, but this isn't possible with Amplify. If prefixes supported wildcards, we could do something like `private/*/original/`, but that's not possible.

This isn't necessarily a bug, just something I wanted to bring up as a caveat. And this is a good example of why I pick and choose which parts of Amplify I use. Auth is great, but I'll likely never use Storage due to limitations like this one.

### PR DESCRIPTION
**Pull request by powerful23**
*Issue #, if available:*
Fixes #611 #332 #668 
*Description of changes:*
Make storage path prefix configurable

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

**Comment by mlabieniec** [comment]
@ffxsam can we get your feedback on this

**Comment by ffxsam** [comment]
Sure! Thanks for asking. 👍 

This could work, I think. So to put a file into the archive, I might do:

```js
Storage.put('documents/howdy.pdf', {
  level: 'private',
  customPrefix: {
    private: 'glacier/private/',
  },
});
```

Then this operation would store the file in `s3://mybucket/glacier/private/<userId>/documents/howdy.pdf`, correct?

**Comment by powerful23** [comment]
@ffxsam yes.

**Comment by alexsmartens** [comment]
Hi,
when are you thinking of merging this pull request?

**Comment by alexsmartens** [comment]
Hi @powerful23 , @ffxsam , @mlabieniec ,
are you planning on merging this pull request?

**Comment by powerful23** [comment]
@mbahar can you check the doc?

**Comment by alexsmartens** [comment]
Would it be possible to release a new Amplify version with this update any time soon?

Thank you for merging the pull request

**Comment by alexsmartens** [comment]
Gentlemen, @powerful23, @richardzcode, @mbahar, @ffxsam, @mlabieniec,
could you please release the new Amplify version?

**Comment by powerful23** [comment]
@alexsmartens yes it's release in aws-amplify@beta for now, we will release an official version next week.

**Comment by alexsmartens** [comment]
I'm still having this problem

I ran `npm install aws-amplify@beta --save`
and updated my storage put method with a custom prefix:
```
 Storage.put( 'FileName', blob, { 
        contentType: 'image/jpeg', 
        level: 'private',
        customPrefix: {
        private: 'MyFolder1/',
      } 
 })
     .then(result => {
      console.log('--result')
     console.log(result)
    })
    .catch(err => {});
```
RESULTS: the files I'm trying to put into the folder at my bucket are not stored anywhere or stored somewhere where I cannot see them

Any ideas?

**Comment by powerful23** [comment]
@alexsmartens can you open the debug mode to see what's the path generated by amplify? ```window.LOG_LEVEL='DEBUG'```

**Comment by alexsmartens** [comment]
Location: "https://MyBucket.s3.us-west-2.amazonaws.com/MyFolder1/SomethingSimilarToIdentityId*/FileName"

SomethingSimilarToIdentityId*:  IdentityId with ':' replaced with '%3A'

**Comment by alexsmartens** [comment]
I also tested put request without the custom prefix configuration.

Here is how the log looks like in this case:
`Location: "https://MyBucket.s3.us-west-2.amazonaws.com/private/SomethingSimilarToIdentityId*/FileName"`

Then, I see this file in this directory on Amazon:
`Amazon S3/MyBucket/private/SomethingSimilarToIdentityId*/FileName `

**Comment by powerful23** [comment]
@alexsmartens did you follow the [doc](https://aws.github.io/aws-amplify/media/storage_guide#Cusomization) to setup your IAM role policy?

**Comment by alexsmartens** [comment]
Yes, here is my policy:

```
"Statement": [
    {
        "Effect": "Allow",
        "Action": [
            "s3:GetObject",
            "s3:PutObject",
            "s3:DeleteObject"
        ],
        "Resource": [
            "arn:aws:s3:::MyBucket/myPublicPrefix/*",
        ]
    }
]
```

**Comment by powerful23** [comment]
So did you put  something like ```"arn:aws:s3:::your-s3-bucket/myPrivatePrefix/${cognito-identity.amazonaws.com:sub}/*"``` to allow you to put object in private folder?

**Comment by alexsmartens** [comment]
I did not have anything like that. I tried adding this policy:
```
{
    "Version": "2012-10-17",
    "Statement": [
        {
            "Effect": "Allow",
            "Action": [
                "s3:GetObject",
                "s3:PutObject",
                "s3:DeleteObject"
            ],
            "Resource": [
                "arn:aws:s3:::MyBucket/myPrivatePrefix/${cognito-identity.amazonaws.com:sub}/*"
            ]
        }
    ]
}
```
where myPrivatePrefix is literally 'myPrivatePrefix'

Adding this policy changed nothing.

**Comment by alexsmartens** [comment]
I also tried running:
```
 Storage.put( 'FileName', blob, { 
        contentType: 'image/jpeg', 
        level: 'public',
        customPrefix: {
        public: 'MyFolder1/',
      } 
 })
     .then(result => {
      console.log('--result')
     console.log(result)
    })
    .catch(err => {});
```

And the location looks like this:
`Location: "https://MyBucket.s3.us-west-2.amazonaws.com/MyFolder1/FileName"`

This is correct location, however I still cannot see my file (FileName) in the bucket (MyBucket)

**Comment by powerful23** [comment]
You need to put your prefix into your policy like:
```
{
    "Version": "2012-10-17",
    "Statement": [
        {
            "Effect": "Allow",
            "Action": [
                "s3:GetObject",
                "s3:PutObject",
                "s3:DeleteObject"
            ],
            "Resource": [
                "arn:aws:s3:::MyBucket/myFolder1/*",
                "arn:aws:s3:::MyBucket/MyFolder1/${cognito-identity.amazonaws.com:sub}/*"
            ]
        }
    ]
}
```

**Comment by alexsmartens** [comment]
Here is my updated code: 
```
{
    "Version": "2012-10-17",
    "Statement": [
        {
            "Effect": "Allow",
            "Action": [
                "s3:GetObject",
                "s3:PutObject",
                "s3:DeleteObject"
            ],
            "Resource": [
                "arn:aws:s3:::MyBucket/MyFolder1/*",
                "arn:aws:s3:::MyBucket/MyFolder1/${cognito-identity.amazonaws.com:sub}/*"
            ]
        }
    ]
}
```

```
 Storage.put( 'FileName', blob, { 
        contentType: 'image/jpeg', 
        level: 'public',
        customPrefix: {
        public: 'MyFolder1/',
      } 
 })
     .then(result => {
      console.log('--result')
     console.log(result)
    })
    .catch(err => {});
```

Log:
`Location: "https://MyBucket.s3.us-west-2.amazonaws.com/MyFolder1/FileName"`

I still cannot see my file in the bucket though

**Comment by alexsmartens** [comment]
We also have a C# aws sdk implementation of the same put method and I noticed that it has a different location structure:
`https://s3.us-west-2.amazonaws.com/MyBucket/MyFolder1/FileName `
opposing to the Amplify implementation:
`https://MyBucket.s3.us-west-2.amazonaws.com/MyFolder1/FileName `

Is is alright?

**Comment by powerful23** [comment]
@alexsmartens Not sure but it could be a problem that the endpoints are different. The endpoints are generated by aws-sdk-js so we don't have any control on it. What's the response of the http request?

**Comment by alexsmartens** [comment]
The response I'm getting is {key: 'FileName'}

**Comment by powerful23** [comment]
@nidsharm can you try out this feature?

**Comment by alexsmartens** [comment]
@mlabieniec, @nidsharm could please help us

**Comment by powerful23** [comment]
@alexsmartens Maybe it's caused by your bucket Name since it has capital letter which might be converted into Non-capital style. Check [Accessing a Bucket](https://docs.aws.amazon.com/AmazonS3/latest/dev/UsingBucket.html) and [Rules for Bucket Naming](https://docs.aws.amazon.com/AmazonS3/latest/dev/BucketRestrictions.html)

File: docs/media/storage_guide.md [hunk 1]
```diff
@@ -228,6 +228,26 @@ Storage.list('photos/', {level: 'private'})
 For the complete API documentation for Storage module, visit our [API Reference]({%if jekyll.environment == 'production'%}{{site.amplify.baseurl}}{%endif%}/api/classes/storageclass.html)
 {: .callout .callout--info}
 
+## Customized Path Prefix
```
Comment (rzubrycki) [inline docs/media/storage_guide.md]: This should be in docs guys!

**Pull request merged**